File: src/Element.cpp
#include "../include/Element.h"
#include <stdexcept>
Element::Element() noexcept {
	this->selected = false;
	this->width = UNDEF;
	this->height = UNDEF;
}
Element::Element(pugi::xml_node& elementNode, std::string elementType) noexcept(false) : 
		matrix(elementNode.child("matrix").child("Matrix"), elementNode), 
		transformationPoint(elementNode.child("transformationPoint").child("Point")) {
	if(std::find(ACCEPTABLE_ELEMENT_TYPES.begin(), ACCEPTABLE_ELEMENT_TYPES.end(), elementType) == ACCEPTABLE_ELEMENT_TYPES.end()) throw std::invalid_argument("Invalid element type: " + elementType);
	this->elementType = elementType;
	this->root = elementNode;
	this->width = UNDEF;
	this->height = UNDEF;
}
Element::Element(const Element& element) noexcept : 
		matrix(element.getMatrix()), 
		transformationPoint(element.getTransformationPoint()) {
	auto parent = element.root.parent();
	this->elementType = element.elementType;
	this->root = parent.insert_copy_after(element.root, element.root);
	this->setWidth(element.getWidth());
	this->setHeight(element.getHeight());
	this->setSelected(element.isSelected());
}
Element::~Element() noexcept {
}
void Element::setWidth(double width) noexcept {
	this->width = width;
}
void Element::setHeight(double height) noexcept {
	this->height = height;
}
const std::string& Element::getElementType() const noexcept {
	return this->elementType;
}
bool Element::isSelected() const noexcept {
	return this->selected;
}
void Element::setSelected(bool selected) noexcept {
	if (!selected) this->root.remove_attribute("isSelected");
	else {
		if (this->root.attribute("isSelected").empty()) this->root.append_attribute("isSelected");
		this->root.attribute("isSelected").set_value(selected);
	}
	this->selected = selected;
}
Matrix& Element::getMatrix() noexcept {
	return this->matrix;
}
const Matrix& Element::getMatrix() const noexcept {
	return this->matrix;
}
void Element::setMatrix(const Matrix& matrix) noexcept {
	this->matrix = matrix;
}
Point& Element::getTransformationPoint() noexcept {
	return this->transformationPoint;
}
const Point& Element::getTransformationPoint() const noexcept {
	return this->transformationPoint;
}
void Element::setTransformationPoint(const Point& transformationPoint) noexcept {
	this->transformationPoint = transformationPoint;
}
pugi::xml_node& Element::getRoot() noexcept {
	return this->root;
}
const pugi::xml_node& Element::getRoot() const noexcept {
	return this->root;
}

File: CsXFL/Element/Element.cs
using System.Xml.Linq;
namespace CsXFL;
public abstract class Element
{
    private static readonly HashSet<string> AcceptableElementTypes = new HashSet<string> { "shape", "text", "tflText", "instance", "shapeObj" };
    internal const string ELEMENTS_NODEGROUP_IDENTIFIER = "elements";
    public static class DefaultValues
    {
        public const double Width = double.NaN;
        public const double Height = double.NaN;
        public const bool Selected = false;
        public const string Name = "";

    }
    protected XElement? root;
    protected XNamespace ns;
    protected string elementType;
    protected string? name;
    protected double width, height;
    protected bool selected;
    protected Matrix matrix;
    protected Point transformationPoint;
    internal XElement? Root { get { return root; } }
    public string ElementType { get { return elementType; } protected set { elementType = value ?? ""; root?.SetAttributeValue("elementType", value); } }
    public string? Name { get { return name; } set { name = value; root?.SetOrRemoveAttribute("name", value, DefaultValues.Name); } }
    public virtual double Width { get { return width; } set { width = value; root?.SetAttributeValue("width", value); } }
    public virtual double Height { get { return height; } set { height = value; root?.SetAttributeValue("height", value); } }
    public bool Selected { get { return selected; } set { selected = value; root?.SetOrRemoveAttribute("isSelected", value, DefaultValues.Selected); } }
    public Matrix Matrix { get { return matrix; } set { SetMatrix(value); } }
    public double ScaleX { get { return matrix.A; } set { matrix.A = value; } }
    public double ScaleY { get { return matrix.D; } set { matrix.D = value; } }
    public Point TransformationPoint { get { return transformationPoint; } }
    internal Element(XNamespace ns)
    {
        root = null;
        this.ns = ns;
        elementType = string.Empty;
        width = double.NaN;
        height = double.NaN;
        selected = false;
        matrix = new Matrix(ns, root);
        transformationPoint = new Point(ns);
    }
    internal Element(in XElement elementNode, string elementType)
    {
        if (!AcceptableElementTypes.Contains(elementType))
        {
            throw new ArgumentException("Invalid element type: " + elementType);
        }
        root = elementNode;
        ns = root.Name.Namespace;
        this.elementType = elementType;
        width = double.NaN;
        height = double.NaN;
        selected = (bool?)elementNode.Attribute("isSelected") ?? DefaultValues.Selected;
        matrix = elementNode.Element(ns + Matrix.MATRIX_NODE_IDENTIFIER)?.Element(ns + Matrix.MATRIX_NODEGROUP_IDENTIFIER) is not null ? new Matrix(elementNode.Element(ns + Matrix.MATRIX_NODE_IDENTIFIER)!.Element(ns + Matrix.MATRIX_NODEGROUP_IDENTIFIER), root) : new Matrix(ns, root);
        transformationPoint = elementNode.Element(ns + Point.TRANSFORMATION_POINT_NODE_IDENTIFIER)?.Element(ns + Point.POINT_NODE_IDENTIFIER) is not null ? new Point(elementNode.Element(ns + Point.TRANSFORMATION_POINT_NODE_IDENTIFIER)!.Element(ns + Point.POINT_NODE_IDENTIFIER)!) : new Point(ns);
    }
    internal Element(in Element other)
    {
        root = other.Root is null ? null : new XElement(other.Root);
        ns = other.ns;
        elementType = other.elementType;
        width = other.width;
        height = other.height;
        selected = other.selected;
        matrix = new Matrix(other.matrix);
        transformationPoint = new Point(other.transformationPoint);
    }
    internal Element(Item item, string elementType, string nodeName) : this(item.Namespace)
    {
        if (!AcceptableElementTypes.Contains(elementType))
        {
            throw new ArgumentException("Invalid element type: " + elementType);
        }
        ns = item.Namespace;
        root = new XElement(ns + nodeName);
        root.Add(new XElement(ns + Point.TRANSFORMATION_POINT_NODE_IDENTIFIER));
        root.Element(ns + Point.TRANSFORMATION_POINT_NODE_IDENTIFIER)?.Add(transformationPoint.Root);
        this.elementType = elementType;
        matrix.SetParent(root);
    }
    private void SetMatrix(Matrix matrix)
    {
        // set values, not the matrix itself
        this.matrix.A = matrix.A;
        this.matrix.B = matrix.B;
        this.matrix.C = matrix.C;
        this.matrix.D = matrix.D;
        this.matrix.Tx = matrix.Tx;
        this.matrix.Ty = matrix.Ty;
    }
}